### ISSUE TITLE
reportProgress not working

### ISSUE THREAD

**Issue by emiryumak** [angular/angular#53650]
### Which @angular/* package(s) are the source of the bug?

common

### Is this a regression?

Yes

### Description

After upgrading my project to Angular 17, I started not being able to listen to reportProgress. I thought it was related to the backend configuration. When I tried jquery ajax at my local, I was able to listen to the events with xhr.upload. The code is below;

	const headers = new HttpHeaders({ 'ngsw-bypass': '' });

	return this.apiService
		.put(
			new ApiCall({
				apiUrl: payload.files.data.upload_to,
				path: '',
				body: payload.files.data.file,
				options: {
					reportProgress: true,
					observe: 'events',
					headers: headers,
				},
			})
		)
		.pipe(
			takeUntil(this.fileFacade.destroy$),
			// map(() => fileAction.uploadFilesSuccess()),
			map((event: HttpEvent<any>) => {
				console.log(event);

				switch (event.type) {
					case HttpEventType.UploadProgress:
						let progress = Math.round((event.loaded / event.total) * 100);
						return fileAction.setUploadProgress({ percent: progress });
					case HttpEventType.Response:
						return fileAction.uploadFilesSuccess();
					default:
						return fileAction.uploadFilesFail(null);
				}
			}),
			catchError((err) => of(fileAction.uploadFilesFail(err)))
		);

When it works this way, only event type 0 is returned, regardless of whether the file is large or small. (HttpEventType.Sent)

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.0.7
Node: 20.9.0
Package Manager: npm 10.1.0
OS: darwin arm64

Angular: 17.0.7
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router, service-worker, ssr

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1700.7
@angular-devkit/build-angular   17.0.7
@angular-devkit/core            17.0.0
@angular-devkit/schematics      17.0.0
@angular/cdk                    17.0.1
@schematics/angular             17.0.7
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2
```

**Comment by JeanMeche** [comment]
Are you using the new `useFetch()`  provider ?

**Comment by emiryumak** [comment]
Yes, i using `withFetch()` provider. When i removed `provideHttpClient(withFetch())` it's working. Can I make it use `XMLHttpRequest` for only a single request?

**Comment by blueiceprj** [comment]
How we can use both HttpClient with Xhr and fetchApi as different injection?

**Comment by JeanMeche** [comment]
You _could_ `provideHttpClient()` without `withFetch()` in a module or a specific component but it is definitly not recommanded.

**Comment by blueiceprj** [comment]
I want to use them in a service (when I need file upload with progress report will use without fetch otherwise use withFetch)

**Comment by imaksp** [comment]
Hello, is there any plan for implementing progress support with fetch, as now with SSR angular shows warning like this if we don't use fetch api:

> NG02801: Angular detected that `HttpClient` is not configured to use `fetch` APIs. It's strongly recommended to enable `fetch` for applications that use Server-Side Rendering for better performance and compatibility. To enable `fetch`, add the `withFetch()` to the `provideHttpClient()` call at the root of the application.
[server]: init 

& If we use fetch upload progress doesn't work.

does this whatwg PR on fetch standard helps?
https://github.com/whatwg/fetch/pull/425

**Comment by JeanMeche** [comment]
Right now Firefox and Safari do not support a `ReadableStream` as a Request body. Until this feature lands, its not possible to provide progress repport with `fetch`.

https://developer.mozilla.org/en-US/docs/Web/API/Request#browser_compatibility

**Comment by DavidGL17** [comment]
Does anyone know if this has been fixed, or if there is a way to solve this ? I'm having the same issue on my project 

Output of ng version : 
```
Angular CLI: 19.1.8
Node: 20.19.1
Package Manager: npm 10.9.0
OS: linux x64

Angular: 19.1.7
... animations, common, compiler, compiler-cli, core, forms
... language-service, localize, platform-browser
... platform-browser-dynamic, platform-server, router

Package                              Version
--------------------------------------------------------------
@angular-devkit/architect            0.1901.8
@angular-devkit/core                 19.1.8
@angular-devkit/schematics           19.1.8
@angular/build                       19.1.8
@angular/cli                         19.1.8
@angular/material-date-fns-adapter   19.2.8
@angular/ssr                         19.1.8
@schematics/angular                  19.1.8
ng-packagr                           19.1.2
rxjs                                 7.8.2
typescript                           5.7.3
zone.js                              0.15.0
```
I'm using it the same way as what is described above, but no upload progress is being reported

**Comment by JeanMeche** [comment]
I'm renaming this issue to illustrate the current limitation.

Note that this limitation is mentionned in the docs : https://v20.angular.dev/api/common/http/withFetch

**Comment by zd-bwxt** [comment]
Could I recommend an option to opt out of fetch and fall back to XHR per-call?  For example:

```ts
this.http.post<AddRequestAttachmentDto>(`api/RequestAttachment/Upload`, formData, {
  reportProgress: true,
  observe: 'events',
  preferXhr: true // <-------- here
});
```

This way we don't have to create an environment-specific injection and re-wire up our interceptors.

**Comment by JeanMeche** [comment]
@zd-bwxt Such API would require that implementation ship both the Xhr and the fetch implementation even if you don't use the Xhr. It's unlikely we'll want to go in that direction.

**Comment by zd-bwxt** [comment]
Could we opt into it some how?  Right now my solution is the following:

Create the XHR provider
```ts
export const XHR_HTTP = new InjectionToken<HttpClient>('XHR_HTTP');

// Minimal XHR factory for browsers
class BrowserXhrFactory implements XhrFactory {
  build(): XMLHttpRequest { return new XMLHttpRequest(); }
}

export function provideXhrHttpClient() {
  return {
    provide: XHR_HTTP,
    deps: [EnvironmentInjector],
    useFactory: (parent: EnvironmentInjector) => {
      const child = createEnvironmentInjector(
        [
          // Build a fresh HttpClient in this child scope…
          provideHttpClient(withInterceptorsFromDi()),
          // …and **force** the transport to XHR here:
          { provide: XhrFactory, useClass: BrowserXhrFactory },
          { provide: HttpBackend, useClass: HttpXhrBackend },
        ],
        parent
      );
      return child.get(HttpClient);
    },
  };
}
```

Globally add both providers
```ts
    // ... add to providers
    provideHttpClient(withInterceptorsFromDi(), withFetch()),
    provideXhrHttpClient(),
```

Add both clients to my service
```ts
@Injectable({
  providedIn: 'root'
})
export class RequestAttachmentService {

  constructor(
    private http: HttpClient,
    @Inject(XHR_HTTP) private xhrHttp: HttpClient,
    ) { }

  // Fetch
  getByRequestId(requestId: number) {
    return this.http.get<RequestAttachmentListDto[]>(`api/RequestAttachment/GetByRequestId/${requestId}`);
  }

  // XHR
  upload(formData: FormData) {
    return this.xhrHttp.post<AddRequestAttachmentDto>(`api/RequestAttachment/Upload`, formData, {
      reportProgress: true,
      observe: 'events',
    });
  }
}
```

Is there any way to simplify all this so we could just reuse `provideHttpClient(withInterceptorsFromDi()),` without the `withFetch()` but register it under the `InjectionToken<HttpClient>('XHR_HTTP');`?
